File: game-server/src/main/java/org/forfun/mmorpg/game/admin/IpUtil.java
package org.forfun.mmorpg.game.admin;

import org.apache.commons.lang3.StringUtils;

import javax.servlet.http.HttpServletRequest;

public class IpUtil {


    private static final String LOCAL_IP = "127.0.0.1";

    public static String getRealIp(HttpServletRequest request) {
        String userName = request.getParameter("userName");
        if (StringUtils.isNotBlank(userName)) {
            return userName;
        }
        String ip = request.getHeader("x-forwarded-for");
        if (ip == null || ip.length() == 0 || "unknown".equalsIgnoreCase(ip)) {
            ip = request.getHeader("Proxy-Client-IP");
        }

        if (ip == null || ip.length() == 0 || "unknown".equalsIgnoreCase(ip)) {
            ip = request.getHeader("WL-Proxy-Client-IP");
        }

        if (ip == null || ip.length() == 0 || "unknown".equalsIgnoreCase(ip)) {
            ip = request.getRemoteAddr();
        }

        if (ip.contains(",")) {
            ip = ip.split(",")[0];
        }

        return "0:0:0:0:0:0:0:1".equals(ip) ? LOCAL_IP : ip;
    }
}

File: game-server/src/main/kotlin/org/forfun/mmorpg/game/admin/IpUtil.kt
package org.forfun.mmorpg.game.admin

import jakarta.servlet.http.HttpServletRequest
import org.apache.commons.lang3.StringUtils
import org.slf4j.LoggerFactory
import java.net.Inet6Address
import java.net.InetAddress
import java.net.InetSocketAddress
import java.net.NetworkInterface
import java.net.SocketAddress

object IpUtil {

    private val LOCAL_IP = "127.0.0.1"
    private val logger = LoggerFactory.getLogger(IpUtil::class.java)

    fun getRealIp(request: HttpServletRequest): String {
        val userName = request.getParameter("userName")
        if (StringUtils.isNotBlank(userName)) {
            return userName!!
        }
        var ip = request.getHeader("x-forwarded-for")
        if (ip.isNullOrEmpty() || "unknown".equals(ip, ignoreCase = true)) {
            ip = request.getHeader("Proxy-Client-IP")
        }

        if (ip.isNullOrEmpty() || "unknown".equals(ip, ignoreCase = true)) {
            ip = request.getHeader("WL-Proxy-Client-IP")
        }

        if (ip.isNullOrEmpty() || "unknown".equals(ip, ignoreCase = true)) {
            ip = request.remoteAddr
        }

        if (ip.contains(",")) {
            ip = ip.split(",")[0]
        }

        return if ("0:0:0:0:0:0:0:1" == ip) LOCAL_IP else ip
    }

}
